### PR TITLE
Suggest restricting to main in GHA publishing guide

### LINKED ISSUE #744: Suggest branch restriction in GitHub Actions tutorial
I was working through the guide at https://packaging.python.org/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/ and made a mistake by creating a workflow that published any time a push is made.

The guide currently suggests:

```yaml
on: push
```

But this means that the release job will run on any push on any branch.

I think instead, we should advise to restrict the job to the master branch:

```yaml
on:
  push:
    branches:
      - master
```

### PR DESCRIPTION
**Pull request by ichard26**
Previously the guide suggested workflow would run on every push which
isn't ideal (eg. a push on for an in-development feature branch doesn't need
a release).

Fixes #744.

### CODE REVIEW

**Comment by ichard26** [comment]
Ugh, made the classic "I'm too used to Markdown" error :facepalm:

File: source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml [hunk 1]
```diff
@@ -1,6 +1,9 @@
 name: Publish Python 🐍 distributions 📦 to PyPI and TestPyPI
 
-on: push
+on:
+  push:
+    branches:
+      - main
```
Comment (webknjaz) [inline source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:6]: I think this change is out of scope. I'd rather not assume the name of the default branch in a random user's repo.

File: source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst [hunk 2]
```diff
@@ -75,7 +80,7 @@ In this guide, we'll use Ubuntu 18.04:
 
 .. literalinclude:: github-actions-ci-cd-sample/publish-to-test-pypi.yml
    :language: yaml
-   :start-after: on:
+   :start-after: - main
```
Comment (webknjaz) [inline source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:83]: It's probably better to do this
```suggestion
   :start-at: jobs:
```

File: source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst [hunk 3]
```diff
@@ -65,6 +65,11 @@ should make GitHub run this workflow:
    :language: yaml
    :end-before: jobs:
 
+The workflow only runs on push events for the ``main`` branch to avoid
+unnecessary publishes, like on a feature branch. If your primary
+branch is named something else, like ``master``, please replace ``main`` with
```
Comment (webknjaz) [inline source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:70]: Do we really want to go into teaching users more about Git/GitHub and default branches? I feel like this may shift the focus from the main goal of this guide.

File: source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml (hunk 1 above)
Comment (webknjaz) [reply source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:6]: Also, this is overly indented and doesn't match the indentation style of the rest of this file. It shouldn't add extra indent in the sequences.

**Review by webknjaz — COMMENTED** [review]
Looking closer at this change, it seems to break the tag publishing support. The assumption was that the tagged releases would go to both TestPyPI and PyPI but the normal pushes would only go to TestPyPI. By only allowing one hardcoded branch, it basically makes the publish step useless.
FTR I've been meaning to rewrite some aspects of this guide and even write a separate one that would be more advanced but never got to it...

**Comment by webknjaz** [comment]
> which
> isn't ideal (eg. a push on for an in-development feature branch doesn't need
> a release).

I think I lean towards having an on-step if-clause. The build dist smoke test should always run. Also, the universal default branch check could look like https://github.com/ansible-community/ansible-pygments/blob/e4830fa/.github/workflows/ci-cd.yml#L155-L157.

**Comment by ichard26** [comment]
Oh dear, it looks like there's a lot of work to do done here to even make this half-acceptable. Thank you for the feedback and my apologies for wasting your time.

Is this even worth pursing? I think by the point I fix the main issues, this would have lost most if not all of its value.

File: source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst (hunk 3 above)
Comment (ichard26) [reply source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst:70]: Yeah, this wasn't a good call. Agreed that's too much of a distraction.

**Comment by webknjaz** [comment]
@ichard26 please don't apologize, you did nothing wrong :) Your contribution is still valuable and seeing how I constantly fail to find time to do the improvements myself, I'd say that we should follow the principle of "better now than never" and try to accept any incremental improvements where possible...

I think it'd be useful to have that if-clause on the "publish to TestPyPI" step in this form:
```yaml
if: >-
  github.event_name == 'push' &&
  github.ref == format(
    'refs/heads/{0}', github.event.repository.default_branch
  )
```
But keep the original "on:" trigger and instead of a long explanation about the branches, I'd just add a `.. tip::` admonition mentioning that the conditional is needed because other branches may trigger this + producing unique per-commit versions would require more work not currently covered in this guide — this is because if the version is hardcoded in a file, different pushes (merges to main) would attempt publishing the same version to TestPyPI effectively failing because of the immutability checks on the server side.

**Comment by ichard26** [comment]
OK, that seems reasonable. Y'know I would expected to \*not\* make this many errors on such a simple issue, but honestly knowing me, this doesn't surprise me. Also, I definitely should've tested the workflow so I understood it better (that probably would've alerted me the fundamental issue with the suggested event change).

Thanks again, I appreciate the encouragement! Let's see if I can get a hold of it on attempt no. 2 ...

P.S. as a maintainer myself, isn't weird how somehow we end pouring more time we think we have when the changes are from someone else? I am acutely aware at how much time can be saved if I just fix an issue myself instead of letting and guiding someone else to do it :)

File: source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml (hunk 1 above)
Comment (ichard26) [reply source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:6]: Not relevant anymore but noted! Thank you!

File: source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml [hunk 4]
```diff
@@ -29,6 +29,11 @@ jobs:
         .
     # Actually publish to PyPI/TestPyPI
     - name: Publish distribution 📦 to Test PyPI
+      if: >-
+        github.event_name == 'push' &&
```
Comment (bhrutledge) [inline source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:33]: Is this redundant, given that the worfklow is already `on: push`?

```suggestion
```

File: source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml [hunk 5]
```diff
@@ -29,6 +29,11 @@ jobs:
...
+        github.ref == format(
+          'refs/heads/{0}', github.event.repository.default_branch
+        )
```
Comment (bhrutledge) [inline source/guides/github-actions-ci-cd-sample/publish-to-test-pypi.yml:36]: @webknjaz I'm not entirely following your earlier suggestions. I think I understand that this will only publish to TestPyPI on a push to `main`. However, it seems like this will still publish to PyPI on any tags of any branch, regardless of whether or not they indicate a release. That doesn't feel desirable to me, and I don't think it addresses the issue reported by @di in https://github.com/pypa/packaging.python.org/issues/744.

Without looking at the GH Actions syntax, could the `if:` be moved to a higher scope, i.e. coupled with the `on:`?

Also, this feels related to https://github.com/pypa/packaging.python.org/issues/777 and https://github.com/pypa/packaging.python.org/issues/804
